### DISCUSSION TITLE
Error 111 Connection refused

### DISCUSSION THREAD

**Discussion by damucka** [bentleymi/ta-openai-api#8] [General]
Hello,

I am getting the following error:

<img width="420" alt="image" src="https://user-images.githubusercontent.com/41383973/215758293-2db9126e-d620-4be5-8c9e-e72be406a10f.png">

I guess this must have sth to do with my port opening. How can I track it further?
What I tried though was the splunk curl command, which we use a lot and really like it .-), but also there I am unable to pass the payload in the correct json form:

<img width="1403" alt="image" src="https://user-images.githubusercontent.com/41383973/215759776-5aa7d4f7-6fb5-479a-a3b1-9dfab88654e4.png">

I tried also eval data = json_object(...) and some other combinations, but still getting same error.
The connect to check models (just header, no data) works fine, so I think the issue must be in the format of the data field.

How would I continue from here?
Best with the ta-openai-api, but if it takes me too many loops with my internal security team for port opening, I would accept also the way over the webtools-add on

Kind Regards,
Kamil

**Comment by bentleymi** [comment]
If | curl doesnt have the same error then this is probably something else (not a firewall / proxy issue).

Did you specify a correct Organization or did you use org-XXXX as pictured?

**Reply by damucka** [reply]
The Organization was correct, taken from the openai settings:

|chatgpt org="org-K5cAcX2Eaq3C6Gk......." prompt="How are you"

Each time when I open the app settings, I get the empty fields for the org ID and API key, but I guess this is for the security reasons not to display them:

![image](https://user-images.githubusercontent.com/41383973/215782912-1b2895ec-b2e8-4a0a-bcd6-bb0f211e7abf.png)

In the passwords.conf I can see the entries for credentials and password, so I think the configuration is correct:

[credential:TA-openai-chatgpt:org-K5cAcX2Eaq3C6Gk.......:]
password = $7$OQEcSGOT1FdP6BG44QpHazwPHu1i..........

Is there any way to get some more insights when digging on the OS level perhaps?
Unfortunately I am not able to execute the script there as well:

_splunk@ccd01v013355:/opt/splunkdev2/splunk/etc/apps/TA-openai-chatgpt/bin> python ./chatgpt.py
  File "./chatgpt.py", line 83
    prompt = (f"How long has Michael Bentley been a member of the Splunk Trust?")
                                                                               ^
SyntaxError: invalid syntax
splunk@ccd01v013355:/opt/splunkdev2/splunk/etc/apps/TA-openai-chatgpt/bin>_

**Reply by damucka** [reply]
I think I should have called it another ways on the OS level. When I do the following, the script is hanging and after cancellation the result is as follows:

_splunk@ccd01v013355:/opt/splunkdev2/splunk/etc/apps/TA-openai-chatgpt/bin> /opt/splunkdev2/splunk/bin/splunk cmd python chatgpt.py  org="org-K5cAcX2Eaq3C6Gk........." prompt="What is your name?"
^CTraceback (most recent call last):
  File "chatgpt.py", line 209, in <module>
    execute()
  File "chatgpt.py", line 148, in execute
    results,dummyresults,settings = splunk.Intersplunk.getOrganizedResults()
  File "/vol1/opt/splunkdev2/splunk/lib/python3.7/site-packages/splunk/Intersplunk.py", line 361, in getOrganizedResults
    results = readResults(input_str, settings)
  File "/vol1/opt/splunkdev2/splunk/lib/python3.7/site-packages/splunk/Intersplunk.py", line 290, in readResults
    line = input_buf.readline()
KeyboardInterrupt
splunk@ccd01v013355:/opt/splunkdev2/splunk/etc/apps/TA-openai-chatgpt/bin>_

or is it totally wrong everything the way I call it?

**Comment by damucka** [comment]
When I test it with curl from my search head on the OS level, I am getting the correct result back:

splunk@ccd01v013355:~> curl https://api.openai.com/v1/completions   -H 'Content-Type: application/json'   -H "Authorization: Bearer sk-U02F66OhGKDEAJChRJMpT3BlbkFJxo.........."   -d '{
  "model": "text-davinci-003",
  "prompt": "What is your name?",
  "max_tokens": 4000,
  "temperature": 1.0

}'
{"id":"cmpl-6f3gQn3Sesbsdk24hLcTP7aY1mgpS","object":"text_completion","created":1675243626,"model":"text-davinci-003","choices":[{"text":"\n\nMy name is Jeff.","index":0,"logprobs":null,"finish_reason":"stop"}],"usage":{"prompt_tokens":5,"completion_tokens":7,"total_tokens":12}}
splunk@ccd01v013355:~>

**Comment by bentleymi** [comment]
The command has to be run within splunk search bar because it streams results from the pipeline through it.  The error you're getting on CLI is telling you it can't find the splunk libraries because of this.

you could try running it with splunk instead

./splunk cmd python /path/to/chatgpt.py

I don't know if it will work that way at all but it's worth a shot.

**Comment by bentleymi** [comment]
More I look at this the more it does feel like a proxy issue.  Do you have a proxy?  Can you ask your network team why you're getting connection refused?
